File: VB/DevExpress.XtraReports.CustomControls/XRDataControlAutoWidthCalculator.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraPrinting

Namespace DevExpress.XtraReports.CustomControls
    Public Class DataControlAutoWidthCalculatorArgs
        Inherits AutoWidthCalculatorArgs

        #Region "Fields"
        Private columns As List(Of XRResizableFieldHeader)
        #End Region

        #Region "Methods"
        Public Sub New(ByVal objects As AutoWidthObjectInfoCollection, ByVal isAutoWidth As Boolean, ByVal maxVisibleWidth As Integer)
            MyBase.New(objects, isAutoWidth, maxVisibleWidth)
        End Sub
        #End Region

        #Region "Properties"
        Public Property VisibleColumns() As List(Of XRResizableFieldHeader)
            Get
                Return Me.columns
            End Get
            Set(ByVal value As List(Of XRResizableFieldHeader))
                Me.columns = value
            End Set
        End Property
        #End Region
    End Class

    'Based on GridViewAutoWidthCalculator
    Friend Class XRDataControlAutoWidthCalculator
        Inherits AutoWidthCalculator

        #Region "Fields"
        Private containerControl As XRTableLikeContainerControl
        #End Region

        #Region "Methods"
        Public Sub New(ByVal control As XRTableLikeContainerControl)
            Me.containerControl = control
        End Sub

        Public Overrides Sub CreateList(ByVal e As AutoWidthCalculatorArgs)
            Dim args As DataControlAutoWidthCalculatorArgs = TryCast(e, DataControlAutoWidthCalculatorArgs)
            Me.Objects.Clear()
            Dim list As IList = args.VisibleColumns
            Dim count As Integer = list.Count
            For i As Integer = 0 To count - 1
                Dim column As XRResizableFieldHeader = DirectCast(list(i), XRResizableFieldHeader)
                Dim minWidth As Integer = CInt((GraphicsUnitConverter.Convert(2, GraphicsDpi.HundredthsOfAnInch, containerControl.Dpi)))
                Dim maxWidth As Integer = CInt((containerControl.WidthF)) - (containerControl.VisibleHeaders.Count - 2) * CInt((GraphicsUnitConverter.Convert(2, GraphicsDpi.HundredthsOfAnInch, containerControl.Dpi)))
                Me.Objects.Add(New AutoWidthObjectInfo(column, minWidth, maxWidth, CInt(column.Width), CInt(column.Width), False))
            Next i
        End Sub

        Protected Overrides Sub DoCalc(ByVal e As AutoWidthCalculatorArgs)
            Dim args As DataControlAutoWidthCalculatorArgs = TryCast(e, DataControlAutoWidthCalculatorArgs)
            Me.CalcAutoWidth(args)
        End Sub

        Protected Overrides Sub DoUpdateRealObject(ByVal info As AutoWidthObjectInfo, ByVal setBothToVisibleWidth As Boolean)
            Dim column As XRResizableFieldHeader = TryCast(info.Obj, XRResizableFieldHeader)
            If column IsNot Nothing Then
                column.Width = info.VisibleWidth
            End If
        End Sub
        #End Region
    End Class


End Namespace

File: CS/DevExpress.XtraReports.CustomControls/XRDataControlAutoWidthCalculator.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraPrinting;

namespace DevExpress.XtraReports.CustomControls
{
    public class DataControlAutoWidthCalculatorArgs : AutoWidthCalculatorArgs
    {
        private List<XRResizableFieldHeader> columns;

        public DataControlAutoWidthCalculatorArgs(AutoWidthObjectInfoCollection objects, bool isAutoWidth, int maxVisibleWidth)
            : base(objects, isAutoWidth, maxVisibleWidth)
        {
        }

        public List<XRResizableFieldHeader> VisibleColumns
        {
            get
            {
                return this.columns;
            }
            set
            {
                this.columns = value;
            }
        }
    }

    //Based on GridViewAutoWidthCalculator
    class XRDataControlAutoWidthCalculator : AutoWidthCalculator
    {
        private XRTableLikeContainerControl containerControl;

        public XRDataControlAutoWidthCalculator(XRTableLikeContainerControl control)
        {
            this.containerControl = control;
        }

        public override void CreateList(AutoWidthCalculatorArgs e)
        {
            DataControlAutoWidthCalculatorArgs args = e as DataControlAutoWidthCalculatorArgs;
            this.Objects.Clear();
            IList list = args.VisibleColumns;
            int count = list.Count;
            for (int i = 0; i < count; i++)
            {
                XRResizableFieldHeader column = (XRResizableFieldHeader)list[i];
                int minWidth = (int)GraphicsUnitConverter.Convert(2, GraphicsDpi.HundredthsOfAnInch, containerControl.Dpi);
                int maxWidth = (int)containerControl.WidthF - (containerControl.VisibleHeaders.Count - 2) * (int)GraphicsUnitConverter.Convert(2, GraphicsDpi.HundredthsOfAnInch, containerControl.Dpi);
                this.Objects.Add(new AutoWidthObjectInfo(column, minWidth, maxWidth, (int)column.Width, (int)column.Width, false));
            }
        }

        protected override void DoCalc(AutoWidthCalculatorArgs e)
        {
            DataControlAutoWidthCalculatorArgs args = e as DataControlAutoWidthCalculatorArgs;
            this.CalcAutoWidth(args);
        }

        protected override void DoUpdateRealObject(AutoWidthObjectInfo info, bool setBothToVisibleWidth)
        {
            XRResizableFieldHeader column = info.Obj as XRResizableFieldHeader;
            if (column != null)
            {
                column.Width = info.VisibleWidth;
            }
        }
    }


}
